### PR TITLE
Implement PartiallySignedTransaction::fee

### PR DESCRIPTION
**Pull request by hashmap**
to calculate fee if previous outputs are available.
Closes https://github.com/rust-bitcoin/rust-bitcoin/issues/1220

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]
ACK 67a900a8d5c6e3e6095cfa87f542bb6c37e6c07c

File: bitcoin/src/util/psbt/error.rs [hunk 1]
```diff
@@ -101,6 +103,7 @@ impl fmt::Display for Error {
             },
             Error::CombineInconsistentKeySources(ref s) => { write!(f, "combine conflict: {}", s) },
             Error::ConsensusEncoding => f.write_str("bitcoin consensus or BIP-174 encoding error"),
+            Error::NegativeFee => f.write_str("negative fee is not allowed"),
```
Comment (Kixunil) [inline bitcoin/src/util/psbt/error.rs:106]: Hmm I think more context like "the PSBT has a negative fee which is not allowed" would be easier to understand.

File: bitcoin/src/util/psbt/mod.rs [hunk 2]
```diff
@@ -437,6 +437,29 @@ impl PartiallySignedTransaction {
...
+    /// ## Errors
+    ///
+    /// The function returns [`Error::MissingUtxo`] when UTXO information for at least one input
+    /// is not present or is invalid.
+    /// If calculated value is negative [`Error::NegativeFee`] is returned.
+    pub fn fee(&self) -> Result<u64, Error> {
+        let mut inputs = 0;
+        for utxo in  self.iter_funding_utxos() {
+            inputs += match utxo {
+                Err(_) => return Err(Error::MissingUtxo),
```
Comment (Kixunil) [inline bitcoin/src/util/psbt/mod.rs:452]: Why ignore the error and why not use `?` ?

File: bitcoin/src/util/psbt/mod.rs [hunk 3]
```diff
@@ -437,6 +437,29 @@ impl PartiallySignedTransaction {
...
+    ///
+    /// ## Errors
+    ///
+    /// The function returns [`Error::MissingUtxo`] when UTXO information for at least one input
+    /// is not present or is invalid.
+    /// If calculated value is negative [`Error::NegativeFee`] is returned.
+    pub fn fee(&self) -> Result<u64, Error> {
+        let mut inputs = 0;
+        for utxo in  self.iter_funding_utxos() {
+            inputs += match utxo {
```
Comment (Kixunil) [inline bitcoin/src/util/psbt/mod.rs:451]: Overflow is not handled. I suggest explicitly checking and returning `Overflow` error.

File: bitcoin/src/util/psbt/mod.rs [hunk 4]
```diff
@@ -437,6 +437,29 @@ impl PartiallySignedTransaction {
...
+                Err(_) => return Err(Error::MissingUtxo),
+                Ok(utxo) => utxo.value,
+            }
+        }
+        let outputs:u64 = self.unsigned_tx.output.iter().map(|out| out.value).sum();
```
Comment (Kixunil) [inline bitcoin/src/util/psbt/mod.rs:456]: Again overflow not handled.

File: bitcoin/src/util/psbt/mod.rs [hunk 5]
```diff
@@ -1689,6 +1712,100 @@ mod tests {
...
+        let mut t2 = t.clone();
+        t2.inputs[0].non_witness_utxo = None;
+        assert!(t2.fee().is_err());
+        // invalid fee
+        let mut t3 = t.clone();
+        t3.unsigned_tx.output[0].value = prev_output_val;
+        assert!(t3.fee().is_err());
+
+    }
+
```
Comment (Kixunil) [inline bitcoin/src/util/psbt/mod.rs:1808]: Please add a test for overflow case when you fix it. This is a very important one.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Thanks! Apart from missing overflow handling (which is a critical bug) it looks quite good.

File: bitcoin/src/util/psbt/error.rs (hunk 1 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/error.rs]: ok, done

File: bitcoin/src/util/psbt/mod.rs (hunk 2 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs]: fixed

File: bitcoin/src/util/psbt/mod.rs (hunk 3 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs]: fixed, added an error variant

File: bitcoin/src/util/psbt/mod.rs (hunk 4 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs]: fixed

File: bitcoin/src/util/psbt/mod.rs (hunk 5 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs:1812]: done

File: bitcoin/src/util/psbt/mod.rs [hunk 6]
```diff
@@ -437,6 +437,26 @@ impl PartiallySignedTransaction {
         // because there has been a new softfork that we do not yet support.
         Err(SignError::UnknownOutputType)
     }
+
+    /// Calculates fee
+    ///
```
Comment (tcharding) [inline bitcoin/src/util/psbt/mod.rs:442]: Possible docs improvement, suggestion only.
```    
    /// Calculates transaction fee.
    ///
    /// 'Fee' being the amount that will be paid for mining a transaction with the current inputs
    /// and outputs i.e., the difference in value of the total inputs and the total outputs.
```

File: bitcoin/src/util/psbt/mod.rs [hunk 7]
```diff
@@ -437,6 +437,26 @@ impl PartiallySignedTransaction {
...
+    /// ## Errors
+    ///
+    /// The function returns [`Error::MissingUtxo`] when UTXO information for at least one input
+    /// is not present or is invalid.
+    /// If calculated value is negative [`Error::NegativeFee`] is returned.
+    /// If an integer overflow happens [`Error::FeeOverflow`] is returned.
```
Comment (tcharding) [inline bitcoin/src/util/psbt/mod.rs:448]: This will be rendered in HTML as a single sentence, perhaps either put it as a single sentence in code
```
    /// The function returns [`Error::MissingUtxo`] when UTXO information for at least one input is
    /// not present or is invalid. If calculated value is negative [`Error::NegativeFee`] is
    /// returned. If an integer overflow happens [`Error::FeeOverflow`] is returned.
```
or as a list if that's preferred
```
    /// - [`Error::MissingUtxo`] when UTXO information for any input is not present or is invalid.
    /// - [`Error::NegativeFee`] if calculated value is negative.
    /// - [`Error::FeeOverflow`] if an integer overflow occurs.
```

In case its useful for you, here is a shell command to build the docs using our custom conditional configuration option `docsrs`:
`RUSTDOCFLAGS="--cfg docsrs" cargo +nightly doc --all-features`

**Review by tcharding — APPROVED** [review]
Thanks for the PR! My review comments are just suggested improvements to docs, the logic of this PR looks good to me.

ACK cfa75a5109f19c2f02c62f1d9fcb589a91c6453d

**Comment by tcharding** [comment]
Your CI fails are:
- Remove `clone` when setting `t4` in tests
- Use `max_value()` instead of `MAX` because `MAX` is not available in Rust 1.41.1 (our current MSRV)

**Review by Kixunil — APPROVED** [review]
ACK cfa75a5109f19c2f02c62f1d9fcb589a91c6453d

Thank you!

File: bitcoin/src/util/psbt/mod.rs (hunk 6 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs]: thanks, updated

File: bitcoin/src/util/psbt/mod.rs (hunk 7 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs]: appreciate your detailed suggestion, updated

**Comment by hashmap** [comment]
> Your CI fails are:
> 
>     * Remove `clone` when setting `t4` in tests
> 
>     * Use `max_value()` instead of `MAX` because `MAX` is not available in Rust 1.41.1 (our current MSRV)

Thanks, fixed. Unfortunately `aarch64-apple-darwin` binaries are available starting from Rust 1.49, will see how to get around it.

**Comment by Kixunil** [comment]
> Unfortunately aarch64-apple-darwin binaries are available starting from Rust 1.49

I assume you're saying you use M1, I think x68_64 binaries should still work even though they would be slightly less efficient.

File: bitcoin/src/util/psbt/mod.rs [hunk 8]
```diff
@@ -437,6 +437,28 @@ impl PartiallySignedTransaction {
...
+    ///
+    /// 'Fee' being the amount that will be paid for mining a transaction with the current inputs
+    /// and outputs i.e., the difference in value of the total inputs and the total outputs.
+    ///
+    /// ## Errors
+    ///
+    /// - [`Error::MissingUtxo`] when UTXO information for any input is not present or is invalid.
+    /// - [`Error::NegativeFee`] if calculated value is negative.
+    /// - [`Error::FeeOverflow`] if an integer overflow occurs.
+    pub fn fee(&self) -> Result<u64, Error> {
```
Comment (Kixunil) [inline bitcoin/src/util/psbt/mod.rs:451]: Oh, wait, sorry for being late I only realized now. We should return `Amount` here.

**Comment by hashmap** [comment]
@Kixunil sorry, I should have been be more specific. On M1 I see 
```
% RUSTUP_TOOLCHAIN=1.41.1 ./contrib/test.sh
+ CRATES=bitcoin
+ for crate in '${CRATES}'
+ cd bitcoin
+ ./contrib/test.sh
+ FEATURES='base64 bitcoinconsensus serde rand secp-recovery'
+ '[' '' = true ']'
+ cargo --version
info: syncing channel updates for '1.41.1-aarch64-apple-darwin'
info: latest update on 2020-02-27, rust version 1.41.1 (f3e1a954d 2020-02-24)
error: target 'aarch64-apple-darwin' not found in channel.  Perhaps check https://doc.rust-lang.org/nightly/rustc/platform-support.html for available targets
```
I'll try to fix it later on.

File: bitcoin/src/util/psbt/mod.rs (hunk 8 above)
Comment (hashmap) [reply bitcoin/src/util/psbt/mod.rs]: done, I left calculations using `u64` because psbt values are `u64` (why btw?)

Comment (Kixunil) [reply bitcoin/src/util/psbt/mod.rs]: Yep, that's correct.

> why

An oversight in API design. We have an issue to change it but nobody did it yet. (We have a bunch of APIs we want to improve so if you're surprised next time it's probably the same reason.)

**Comment by Kixunil** [comment]
I think this could work: `rustup toolchain install 1.41.1-x86_64-apple-darwin`
You probably need to run it with `cargo +1.41.1 test --target x86_64-apple-darwin` though, not sure.

**Review by Kixunil — APPROVED** [review]
ACK c34d5f8f859c151df77bbdef7e0ef212a80ed69d if CI passes

**Review by tcharding — APPROVED** [review]
ACK c34d5f8f859c151df77bbdef7e0ef212a80ed69d

**Review by apoelstra — APPROVED** [review]
ACK c34d5f8f859c151df77bbdef7e0ef212a80ed69d
